core/drivers/crypto/hisilicon/sec_main.c

```diff
@@ -0,0 +1,266 @@
+	io_write32(qm->io_base + SEC_RAS_NFE_ENABLE, SEC_RAS_NFE_ENB_MASK);
+	io_write32(qm->io_base + SEC_CORE_INT_MASK, SEC_ABNML_INT_DISABLE);
+
+	sec_enable_clock_gate(qm);
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static enum hisi_drv_status sec_dev_status_check(struct hisi_qm *qm)
+{
+	uint32_t val;
+
+	val = io_read32(qm->io_base + SEC_CORE_INT_SOURCE);
+	if (val & SEC_RAS_NFE_ENB_MASK) {
+		EMSG("SEC NFE RAS happened, need to reset.");
+		return HISI_QM_DRVCRYPT_HW_EACCESS;
+	}
+
+	val = io_read32(qm->io_base + HISI_QM_ABNML_INT_SRC);
+	if (val) {
+		if (val & HISI_QM_SEC_NFE_INT_MASK)
+			EMSG("QM NFE RAS happened, need to reset.");
+
+		if (val & HISI_QM_INVALID_DB) {
+			EMSG("QM invalid db happened, please check.");
+			io_write32(qm->io_base + HISI_QM_ABNML_INT_SRC,
+				   HISI_QM_INVALID_DB);
+		}
+
+		return HISI_QM_DRVCRYPT_HW_EACCESS;
+	}
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static enum hisi_drv_status sec_qm_init(struct acc_device *sec_dev)
+{
+	struct hisi_qm *qm = &sec_dev->qm;
+
+	if (cpu_mmu_enabled()) {
+		qm->io_base = (uintptr_t)phys_to_virt_io(sec_dev->io_base,
+							 sec_dev->io_size);
+		if (!qm->io_base) {
+			EMSG("Fail to get qm io_base!");
+			return HISI_QM_DRVCRYPT_EFAULT;
+		}
+	} else {
+		qm->io_base = sec_dev->io_base;
+	}
+
+	qm->fun_type = sec_dev->fun_type;
+	qm->vfs_num = sec_dev->vfs_num;
+	qm->sqe_size = SEC_SQE_SIZE;
+	qm->sqe_log2_size = SEC_SQE_LOG2_SIZE;
+	if (qm->fun_type == HISI_QM_HW_PF) {
+		hisi_qm_get_version(qm);
+		IMSG("SEC hardware version is %#"PRIx32, qm->version);
+		qm->qp_base = HISI_QM_PF_Q_BASE;
+		qm->qp_num = HISI_QM_PF_Q_NUM;
+		qm->dev_status_check = sec_dev_status_check;
+	}
+
+	return hisi_qm_init(qm);
+}
+
+static struct acc_device *sec_pre_init(void)
+{
+	struct acc_device *sec_dev = NULL;
+
+	sec_dev = calloc(1, sizeof(*sec_dev));
+	if (!sec_dev) {
+		EMSG("Fail to alloc sec_dev!");
+		return NULL;
+	}
+
+	sec_dev->io_base = SEC_BAR;
+	sec_dev->io_size = SEC_SIZE;
+	SLIST_INSERT_HEAD(&sec_list, sec_dev, link);
+
+	return sec_dev;
+}
+
+static TEE_Result sec_probe(void)
+{
+	enum hisi_drv_status ret = HISI_QM_DRVCRYPT_NO_ERR;
+	struct acc_device *sec_dev = NULL;
+	struct hisi_qm *qm = NULL;
+
+	IMSG("SEC driver init start, version %s!", SEC_MODULE_VERSION);
+	sec_dev = sec_pre_init();
+	if (!sec_dev)
+		return HISI_QM_DRVCRYPT_ENOMEM;
+
+	qm = &sec_dev->qm;
+	ret = sec_qm_init(sec_dev);
+	if (ret) {
+		EMSG("Fail to init sec qm, ret=%d", ret);
+		goto err_with_pre_init;
+	}
+
+	ret = sec_engine_init(sec_dev);
+	if (ret) {
+		EMSG("fail to init engine, ret=%d", ret);
+		goto err_with_qm_init;
+	}
+
+	ret = hisi_qm_start(qm);
+	if (ret) {
+		EMSG("Fail to start qm, ret=%d", ret);
+		goto err_with_qm_init;
+	}
+
+	IMSG("SEC driver init done!");
+	return HISI_QM_DRVCRYPT_NO_ERR;
+
+err_with_qm_init:
+	hisi_qm_uninit(qm);
+err_with_pre_init:
+	SLIST_REMOVE_HEAD(&sec_list, link);
+	free(sec_dev);
+
+	return HISI_QM_DRVCRYPT_FAIL;
```

Comment:
please fix the error codes returned by this function.

Reply:
I don't quite understand your comment, how to fix the error codes?

Reply:
This is a `driver_init()` initcall function. It should return a `TEE_Result` code. `HISI_QM_DRVCRYPT_FAIL` and `HISI_QM_DRVCRYPT_NO_ERR` are not `TEE_Result` value (even if `HISI_QM_DRVCRYPT_NO_ERR` is 0 as `TEE_SUCCESS` is).